infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java

```diff
@@ -477,54 +536,168 @@ private void analyzePollReportAndTriggerHandler(PollReport pollReport) {
-                        localManagementClient.handleFailure(pollReport.getFailingNodes()).get();
-                        return;
-                    }
-                }
-                log.debug("Failure already taken care of.");
+            Set<String> failedNodes = new HashSet<>();
+            Set<String> healedNodes = new HashSet<>();
 
-            } else {
-                // CASE 3:
-                // Failures detected but not marked in the layout or
-                // some servers have been partially sealed to new epoch or stuck on
-                // the previous epoch.
-                localManagementClient.handleFailure(pollReport.getFailingNodes()).get();
-                // TODO: Only re-sealing needed if server stuck on a previous epoch.
+            healedNodes.addAll(getHealedNodes(pollReport));
+            failedNodes.addAll(getNewFailures(pollReport));
+
+            if (failedNodes.isEmpty() && healedNodes.isEmpty()) {
+                if (checkIfCurrentLayoutSlotUnFilled(pollReport)) {
+                    log.info("Current layout slot is empty. Filling slot with current layout.");
+                    localManagementClient
+                            .handleFailure(Collections.emptySet(), Collections.emptySet()).get();
+                }
+                return;
             }
 
+            log.info("Detected changes in node responsiveness: Failed:{}, Healed:{}, pollReport:{}",
+                    failedNodes, healedNodes, pollReport);
+            localManagementClient.handleFailure(failedNodes, healedNodes).get();
+
         } catch (Exception e) {
             log.error("Exception invoking failure handler : {}", e);
         }
     }
 
+    /**
+     * Corrects out of phase epochs by resealing the servers.
+     * This would also need to update trailing layout servers.
+     *
+     * @param pollReport Poll Report from running the failure detection policy.
+     */
+    private void correctOutOfPhaseEpochs(PollReport pollReport) {
+        final Map<String, Long> outOfPhaseEpochNodes = pollReport.getOutOfPhaseEpochNodes();
+        if (outOfPhaseEpochNodes.isEmpty()) {
+            return;
+        }
+
+        try {
+            // Query all layout servers to get quorum Layout.
+            Map<String, CompletableFuture<Layout>> layoutCompletableFutureMap = new HashMap<>();
+            for (String layoutServer : latestLayout.getLayoutServers()) {
+                layoutCompletableFutureMap.put(
+                        layoutServer, getCorfuRuntime().getRouter(layoutServer)
+                                .getClient(LayoutClient.class).getLayout());
+            }
+
+            // Retrieve the correct layout from quorum of members to reseal servers.
+            // If we are unable to reach a consensus from a quorum we get an exception and abort the
+            // epoch correction phase.
+            safeUpdateLayout(fetchQuorumLayout(layoutCompletableFutureMap.values()
```

Comment:
Moved this quorumLayoutFetch to the top.
Now every polling interval (1 sec) the management server does a quorum fetch for the layout for which it polls ever layout server and then calculates the layout agreed upon by the majority.
This could be an issue down the line until the poll/heartbeat message itself also carries the layout response from the node.